src/compiler/checker.ts

```diff
@@ -14406,12 +14437,113 @@ namespace ts {
+                                error(local.valueDeclaration, Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                            else if (local.declarations &&
+                                (local.declarations[0].kind === SyntaxKind.InterfaceDeclaration || local.declarations[0].kind === SyntaxKind.ModuleDeclaration)) {
+                                error(local.declarations[0], Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                            else if (local.declarations && local.declarations[0].kind === SyntaxKind.ImportEqualsDeclaration) {
+                                error(local.declarations[0].name, Diagnostics._0_is_declared_but_never_used, key);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        function checkUnusedClassLocals(node: ClassDeclaration): void {
+            if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
+                if (node.members) {
+                    for (const member of node.members) {
+                        if (member.kind === SyntaxKind.MethodDeclaration || member.kind === SyntaxKind.PropertyDeclaration) {
+                            if (isPrivateClassElement(member) && !member.symbol.hasReference) {
+                                error(member, Diagnostics._0_is_declared_but_never_used, member.symbol.name);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        function checkUnusedTypeParameters(node: ClassDeclaration | FunctionDeclaration | MethodDeclaration | FunctionExpression | ArrowFunction | ConstructorDeclaration | SignatureDeclaration) {
+            if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
+                if (node.typeParameters) {
+                    for (const typeParameter of node.typeParameters) {
+                        if (!typeParameter.symbol.hasReference) {
+                            error(typeParameter, Diagnostics._0_is_declared_but_never_used, typeParameter.symbol.name);
+                        }
+                    }
+                }
+            }
+        }
+
+        function isPrivateClassElement(node: ClassElement): boolean {
+            return (node.flags & NodeFlags.Private) !== 0;
+        }
+
+        function checkUnusedImports(node: SourceFile) {
+            if (compilerOptions.noUnusedLocals && !isInAmbientContext(node)) {
+                for (const local in node.locals) {
+                    if (hasProperty(node.locals, local)) {
+                        const localValue = node.locals[local];
+                        if (localValue.declarations && !localValue.exportSymbol) {
+                            for (const declaration of localValue.declarations) {
+                                const symbol = declaration.symbol;

```

Comment:
i do not think this is correct. you already have the symbol, and the symbol is not exported, so `localValue === localValue.declarations[0].symbol`.